hudi-common/src/test/java/org/apache/hudi/common/table/read/TestKeyBasedFileGroupRecordBuffer.java

```diff
@@ -0,0 +1,322 @@
+  }
+
+  private static List<IndexedRecord> getActualRecords(FileGroupRecordBuffer<IndexedRecord> fileGroupRecordBuffer) throws IOException {
+    List<IndexedRecord> actualRecords = new ArrayList<>();
+    while (fileGroupRecordBuffer.hasNext()) {
+      actualRecords.add(fileGroupRecordBuffer.next());
+    }
+    return actualRecords;
+  }
+
+  /**
+   * A custom payload implementation for testing purposes that marks records as deleted once the counter exceeds 2.
+   * During the merge, it will combine the counter values.
+   */
+  public static class CustomPayload extends BaseAvroPayload
+      implements HoodieRecordPayload<CustomPayload> {
+    private final GenericRecord payloadRecord;
+
+    public CustomPayload(GenericRecord record, Comparable orderingVal) {
+      super(record, orderingVal);
+      this.payloadRecord = record;
+    }
+
+    @Override
+    public CustomPayload preCombine(CustomPayload oldValue) {
+      return this;
+    }
+
+    @Override
+    public Option<IndexedRecord> combineAndGetUpdateValue(IndexedRecord currentValue, Schema schema) throws IOException {
+      if (currentValue.get(2).equals(payloadRecord.get(2))) {
+        // If the timestamps are the same, we do not update
+        return Option.of(currentValue);
+      }
+      int result = (int) currentValue.get(1) + (int) payloadRecord.get(1);
+      if (result > 2) {
+        return Option.empty();
+      }
+      return Option.of(createTestRecord(currentValue.get(0).toString(), result, (long) payloadRecord.get(2)));
+    }
+
+    @Override
+    public Option<IndexedRecord> getInsertValue(Schema schema) throws IOException {
+      return Option.of(payloadRecord);
+    }
+
+    @Override
+    public Comparable<?> getOrderingValue() {
+      return null;
+    }
+  }
+
+  public static class CustomMerger implements HoodieRecordMerger {
+
+    @Override
+    public Option<Pair<HoodieRecord, Schema>> merge(HoodieRecord older, Schema oldSchema, HoodieRecord newer, Schema newSchema, TypedProperties props) throws IOException {
+      GenericRecord olderData = (GenericRecord) older.getData();
+      GenericRecord newerData = (GenericRecord) newer.getData();
+      if (olderData.get(2).equals(newerData.get(2))) {
+        // If the timestamps are the same, we do not update
+        return Option.of(Pair.of(older, oldSchema));
+      }
+      int result = (int) olderData.get(1) + (int) newerData.get(1);
+      if (result > 2) {
+        return Option.empty();
+      }
+      HoodieKey hoodieKey = older.getKey();
+      return Option.of(Pair.of(new HoodieAvroIndexedRecord(createTestRecord(hoodieKey.getRecordKey(), result, (long) newerData.get(2))), SCHEMA));
+    }
+
+    @Override
+    public HoodieRecord.HoodieRecordType getRecordType() {
+      return HoodieRecord.HoodieRecordType.AVRO;
+    }
+
+    @Override
+    public String getMergingStrategy() {
+      return "";
```

Comment:
nit: for custom merger, the merger implementation class should be fetched by looking up the merging strategy ID.  To avoid confusion, we could return a random UUID here in the test.